lib/jsdom/living/events/EventTarget-impl.js

```diff
@@ -82,63 +87,134 @@ class EventTargetImpl {
     return this._dispatch(eventImpl);
   }
 
+  // https://dom.spec.whatwg.org/#concept-event-dispatch
   _dispatch(eventImpl, targetOverride) {
+    let targetImpl = this;
+    let clearTargets = false;
+    let activationTarget = null;
+
     eventImpl._dispatchFlag = true;
-    eventImpl.target = targetOverride || this;
 
-    const eventPath = [];
-    let { target } = eventImpl;
-    let targetParent = domSymbolTree.parent(target);
-    while (targetParent) {
-      eventPath.push(targetParent);
-      target = targetParent;
-      targetParent = domSymbolTree.parent(targetParent);
-    }
-    if (eventImpl.type !== "load" && target._defaultView) {
-      // https://html.spec.whatwg.org/#events-and-the-window-object
-      eventPath.push(idlUtils.implForWrapper(target._defaultView));
-    }
+    targetOverride = targetOverride || targetImpl;
+    let relatedTarget = retarget(eventImpl.relatedTarget, targetImpl);
 
-    eventImpl.eventPhase = Event.CAPTURING_PHASE;
-    for (let i = eventPath.length - 1; i >= 0; --i) {
-      if (eventImpl._stopPropagationFlag) {
-        break;
+    if (targetImpl !== relatedTarget || targetImpl === eventImpl.relatedTarget) {
+      const touchTargets = [];
+
+      appendToEventPath(eventImpl, targetImpl, targetOverride, relatedTarget, touchTargets, false);
+
+      const isActivationEvent = MouseEvent.isImpl(eventImpl) && eventImpl.type === "click";
+      if (isActivationEvent) {
+        activationTarget = targetImpl;
       }
 
-      const object = eventPath[i];
-      const objectImpl = idlUtils.implForWrapper(object) || object; // window :(
-      const eventListeners = objectImpl._eventListeners[eventImpl.type];
-      invokeEventListeners(eventListeners, object, eventImpl);
-    }
+      let slotInClosedTree = false;
+      let slotable = isSlotable(targetImpl) && targetImpl._assignedSlot ? targetImpl : null;
+      let parent = getParent(targetImpl, eventImpl);
+
+      // Populate event path
+      // https://dom.spec.whatwg.org/#event-path
+      while (parent !== null) {
+        if (slotable !== null) {
+          if (parent.localName !== "slot") {
+            throw new Error(`JSDOM Internal Error: Expected parent to be a Slot`);
+          }
+
+          slotable = null;
+
+          const parentRoot = getRoot(parent);
+          if (isShadowRoot(parentRoot) && parentRoot.mode === "closed") {
+            slotInClosedTree = true;
+          }
+        }
+
+        if (isSlotable(parent) && parent._assignedSlot) {
+          slotable = parent;
+        }
+
+        relatedTarget = retarget(eventImpl.relatedTarget, parent);
+
+        // TODO: Understand why checking for Window is needed? Window is not a Node...
+        if (
+          (isNode(parent) && isShadowInclusiveAncestor(getRoot(targetImpl), parent)) ||
+          idlUtils.wrapperForImpl(parent).constructor.name === "Window"
```

Comment:
In order to properly dispatch event on `Window`, we also need to check if the *parent* is `Window` since `Window` is not a `Node`. This is not consistent with the [dispatch event](https://dom.spec.whatwg.org/#ref-for-boundary-point-node%E2%91%A0) algorithm

Reply:
I assume tests fail if you follow the spec exactly? In that case, maybe file a spec bug?